core/app/models/spree/payment/processing.rb

```diff
@@ -35,12 +35,37 @@ def process!
       end
 
       def authorize!
-        handle_payment_preconditions { process_authorization }
+        return unless check_payment_preconditions!
+
+        started_processing!
+
+        protect_from_connection_error do
+          response = payment_method.authorize(
+            money.money.cents,
+            source,
+            gateway_options,
+          )
+          pend! if handle_response(response)
+        end
       end
 
       # Captures the entire amount of a payment.
       def purchase!
-        handle_payment_preconditions { process_purchase }
+        return unless check_payment_preconditions!
+
+        started_processing!
+
+        protect_from_connection_error do
+          response = payment_method.purchase(
+            money.money.cents,
+            source,
+            gateway_options,
+          )
+          complete! if handle_response(response)
+        end
+
+        # This won't be called if `payment_method.purchase` raises a GatewayError
```

Comment:
Reading it again now, is this comment needed? I mean I don't expect this line to be executed if the method called before this line raises an exception. WDYT?

Reply:
@elia please one last feedback here and we are ready to merge!

Reply:
Good point, ruby 1:1 😅